File: python/39. Combination Sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        sorted(candidates)
        
        res = []
        def dfs(idx, cur, a):
            if idx >= len(candidates):
                if cur==target:
                    res.append(a)      
                return
          
            k = 0
            while cur+k*candidates[idx] <= target:
                dfs(idx+1, cur+k*candidates[idx], a + k*[candidates[idx]])
                k += 1
                
        a = []
        dfs(0, 0, a)
        return res

        
        
s = Solution()
print s.combinationSum([2,3,6,7], 7)

File: go/src/典型问题/组合/39. Combination Sum/39. Combination Sum.go
package main

import "fmt"

/**
https://leetcode-cn.com/problems/combination-sum/
 */
func combinationSum(candidates []int, target int) [][]int {
	n := len(candidates)

	var ans = [][]int{}
	var backtrace func(int, int, []int)
	backtrace = func(idx int, s int, arr []int) {
		if s <= 0|| idx == n {
			if s == 0 {
				ans = append(ans, append([]int{}, arr...))
			}
			return
		}

		for i := idx; i < n; i++ {
			arr = append(arr, candidates[i])
			backtrace(i, s - candidates[i], arr)	// 可以选多次
			arr = arr[:len(arr) - 1]
		}
	}

	backtrace(0, target, []int{})
	fmt.Println(ans)
	return ans
}

func main() {
	combinationSum([]int{2,3,6,7}, 7)
}
